go/roothash/api/commitment/pool.go

```diff
@@ -365,201 +344,101 @@ func (p *Pool) ProcessCommitments(didTimeout bool) (OpenCommitment, error) {
 		}
 
 		total++
-		commit, ok := p.ExecuteCommitments[n.PublicKey]
-		if !ok {
+		vote, ok := sc.Votes[n.PublicKey]
+		switch {
+		case !ok:
 			continue
+		case vote == nil:
+			failures++
+		default:
+			votes[*vote]++
 		}
 		commits++
 
-		if commit.IsIndicatingFailure() {
-			failures++
+		// Early discrepancy detection.
+		switch {
+		case p.Discrepancy:
+			// Discrepancy resolution already started.
+			continue
+		case len(votes) <= 1 && failures <= int(allowedStragglers):
+			// Discrepancy not detected.
+			continue
+		case p.HighestRank > 0 && !timeout:
+			// To ensure fairness, the backup schedulers are always required to wait for a round
+			// timeout. Without this delay, there is a risk that they could bypass the scheduling
+			// order by convincing a few other nodes to instantly publish failures or discrepant
+			// commitments.
+			//
+			// In the former scenario, the colliding nodes might avoid punishment because nodes
+			// posting a failure are merely marked as 'not alive' for the current round, and no
+			// slashing occurs. In the latter case, the colliding nodes may be penalized, but
+			// the potential gains from being a scheduler could outweigh the losses.
 			continue
 		}
 
-		k := commit.ToVote()
-		if v, ok := votes[k]; !ok {
-			votes[k] = &vote{
-				commit: commit,
-				tally:  1,
-			}
-		} else {
-			v.tally++
-		}
-
-		// As soon as there is a discrepancy we can proceed with discrepancy resolution.
-		// No need to wait for all commits.
-		if !p.Discrepancy && len(votes) > 1 {
-			p.Discrepancy = true
-			return nil, ErrDiscrepancyDetected
-		}
-	}
-
-	// Determine whether the proposer has submitted a commitment.
-	proposer, err := p.Committee.TransactionScheduler(p.Round)
-	if err != nil {
-		return nil, ErrNoCommittee
-	}
-	proposerCommit, ok := p.ExecuteCommitments[proposer.PublicKey]
-	if !ok && didTimeout {
-		// TODO: Consider slashing for this offense.
-		return nil, ErrNoProposerCommitment
+		// The committee is neither unanimous nor is the number of valid commitments expected to
+		// surpass the required threshold due to numerous failures. Start discrepancy resolution.
+		return nil, ErrDiscrepancyDetected
 	}
 
 	switch p.Discrepancy {
 	case false:
 		// Discrepancy detection.
-		allowedStragglers := int(p.Runtime.Executor.AllowedStragglers)
-
-		// If it is already known that the number of valid commitments will not exceed the required
-		// threshold, there is no need to wait for the timer to expire. Instead, proceed directly to
-		// the discrepancy resolution mode, regardless of any additional commits.
-		if failures > allowedStragglers {
-			p.Discrepancy = true
-			return nil, ErrDiscrepancyDetected
+		if len(votes) > 1 || failures > int(allowedStragglers) {
+			// Backup schedulers need to wait for a round timeout.
+			return nil, ErrStillWaiting
```

Comment:
This is because the `"Early discrepancy detection"` would already catch the discrepancy if `p.HighestRank=0` right, so we know that this means that the primary scheduler has not yet committed?

Maybe include this in a comment (all the facts already assumed here due to the early discrepancy checks and not rechecked), as this was not immediately clear to me.

Reply:
> This is because the "Early discrepancy detection" would already catch the discrepancy if p.HighestRank=0 right. 

Yes. If the highest rank is 0, this means that the primary scheduler already committed and the early discrepancy detection handled this case. Therefore, this `if` will never execute.

, so we know that this means that the primary scheduler has not yet committed?
 and can only be true for backup schedulers (those with `rank > 0`).

Reply:
> This is because the `"Early discrepancy detection"` would already catch the discrepancy if `p.HighestRank=0` right.

Yes. If the highest rank is 0, this means that the primary scheduler already committed and the early discrepancy detection handled this case. Therefore, this `if` cannot happen.

> so we know that this means that the primary scheduler has not yet committed

Yes. If this `if` is true, we know that the primary scheduler hasn't committed but some other backup scheduler has. And backup schedulers need to wait for round timeout, as mentioned above.

Will update the comment.